### PR TITLE
__package_pip: fix extras state detection

### LINKED ISSUE #34: __package_pip: undo `--extras` deprecation
In `__package_pip`, the parameter `--extras` was deprecated in #31 cause of unifying parameters:
> > deprecate --extra?
>
> When we have one --requirement to rule them all I think it would only be consequent to do so.

After writing writing with @4nd3r about this in [#cdist:ungleich.ch](https://matrix.to/#/#cdist:ungleich.ch) (around 12:58-13:52 GMT+2), we came to the conclusion, that at least some of the parts are required that the type can function as same as before.

The extras mechanism currently does the following:
1. Checks via `explorer/extras` which given `--extras` aren't currently installed and outputs all extras, that aren't fully installed
2. Installs the extras from the explorers (that currently aren't fully installed)

Point 1 is required that extras can be added after the package is fully installed. Removing is out of scope, as pip don't support the cleanup of unreferenced packages (like `apt autoremove`). Point 2 is the cherry on the cake, but completely useless as it's no different for pip if all extras are directly passed to pip instead of the extra filter for completely installed extras.

The newly introduced `--requirement`, which doesn't completely replace the deprecated `--name` (see `explorer/state` at least), should also be checked and parsed. Because it's also possible to pass extras with it, it should also be parsed in `explorer/state` to benefit of the explorer.

Possible things to do:
- [ ] undo `--extras` deprecation
- [ ] restore `--extras` examples in man
- [ ] support, discard or keep the current state of extras in `--requirement`
- [ ] rework `explorer/extras` comments if I get to touch it for the changes

### PR DESCRIPTION
**Pull request by sideeffect42**
This PR should fix #34.

Please give it a good test, @4nd3r @Matze1224.
I did some tests, but don't use any Python libraries with extras in "production" that I could test the changes with.

### CODE REVIEW

**Comment by 4nd3r** [comment]
Lots of changes, gotta try it out too, since I use this type a lot.

@Matze1224 can you please also give it a go?

**Comment by 4nd3r** [comment]
Probably should be done in separate PR, but...

```
gencode-remote:stderr
---------------------
Could not find pip.
Please make sure that pip is installed or specify the command to use in --pip.
```

Do we want to install `pip` in type manifest?

**Comment by sideeffect42** [comment]
> Do we want to install `pip` in type manifest?

I thought about it, but then decided against it because pip is coupled to the Python version/implementation (PyPy?). It isn't unlikely to have multiple versions of Python installed in parallel, so I concluded that it makes more sense that the user decides which pip they want to use.

**Comment by 4nd3r** [comment]
Good reasoning. Let's add that to manual.

File: type/__package_pip/gencode-remote [hunk 1]
```diff
@@ -19,76 +17,47 @@
...
-        echo "removed" >> "$__messages_out"
-    ;;
-    *)
-        echo "Unknown state: $state_should" >&2
-        exit 1
-    ;;
+case ${state_should}
+in
+	(present)
+		do_cmd "pip install -q $(quote "${requirement}")"
```
Comment (4nd3r) [inline type/__package_pip/gencode-remote:52]: `..."$pip ...`

and same for other states

Comment (sideeffect42) [reply type/__package_pip/gencode-remote]: Obviously 🤦, thanks.

File: type/__package_pip/man.rst [hunk 2]
```diff
@@ -9,7 +9,15 @@ cdist-type__package_pip - Manage packages with pip
...
-It is also included in the python virtualenv environment.
+It is also included in the Python virtualenv environment.
+
+**NB:** This type requires that pip is already installed on the target host.
+Since pip is tightly coupled to a specific Python interpreter and it is common
+to have multiple interpreters (e.g. CPython 2.x, 3.x and/or PyPy) and
+venvs (virtual environments) installed at the same time, this type does not try
+to guess which pip to use.
+Most OSes provide a package to get a default pip installation, so it's usually
+sufficient to add ``__package python3-pip`` to your manifest.
```
Comment (4nd3r) [inline type/__package_pip/man.rst:20]: `python3-pip` yes, but let's add something like "Installable package might have additional dependencies (e.g. some `-dev` package) which have to be installed from distro repository before calling __package_pip"...

Comment (sideeffect42) [reply type/__package_pip/man.rst:20]: What do you mean with "some `-dev` package"?
Pip packages that don't have a wheel for your platform and you need to compile native code directly on the target?

If so, I think we'd be going too far.
Some distros need `-dev` packages (or whatever the suffix is), some don't, on others you compile things from source anyway so you already have the headers.
And if you need to compile native code, you'll also need a compiler toolchain installed. That's usually done by installing `build-essential`/`build-base` or whatever, but then things aren't limited to C. Some people (looking at you, cryptography) want to bundle Rust code, so you need a Rust toolchain, too.

What I'm trying to say is, things are complicated and I think that compressing it down to a paragraph in the man page is not going to work.
Our users are admins, they should be able to handle it.

Comment (sideeffect42) [reply type/__package_pip/man.rst:20]: We could add a sentence on `python3-wheel`, maybe.

**Review by 4nd3r — APPROVED** [review]

**Pull request merged**